### PR TITLE
contractcourt+sweep: fix fee function and deadline issue

### LINKED ISSUE #8738: [bug]: contractcourt: max budget used immediately for anchor sweeps
For outgoing HTLCs the commitment deadline is calculated incorrectly, causing the entire anchor budget to be spent immediately.

### Problem

The commitment deadline is derived from the timelocks of the outgoing HTLCs, instead of the timelocks for the matching upstream (incoming) HTLCs.

https://github.com/lightningnetwork/lnd/blob/c68778d2f39f13617e05b362fb9bbec43e5c7a43/contractcourt/channel_arbitrator.go#L1452-L1480

Since `ChannelArbitrator` only broadcast commitments for outgoing HTLCs once the timelock has already expired, the resulting deadline for anchor sweeping is always 0 (which is bumped to 1).  This causes the sweeper to spend the *entire* budget immediately.

### Solution

The upstream HTLC timelocks need to be used for the deadline instead.

### LINKED ISSUE #8741: [bug]: sweep: LinearFeeFunction off by one
`LinearFeeFunction` doesn't return the max fee rate until *after* the deadline has been missed.

### Example

Current block height is `H`.  An incoming HTLC has a `cltv_expiry` of `H + 10`, so the contractcourt sets `deadline := H + 10` for the HTLC input.

We **need** to confirm our HTLC-Preimage no later than block `H + 10`, or else we'll get into a bidding war with the channel peer since they can spend the timeout path after block `H + 10` confirms.

`LinearFeeFunction` *should* max out its fee rate after block `H + 9` confirms since this is the last shot at meeting the deadline.  Instead, `LinearFeeFunction` waits to max out the fee rate until block `H + 10`.

### Severity

For typical deadlines 10+ blocks away, this is a minor problem.  But in the extreme case of a next-block deadline, `LinearFeeFunction` will return the min feerate instead of the max, which is quite bad.

### PR DESCRIPTION
**Pull request by yyforyongyu**
Fixes #8741 
Fixes #8738

TODO:
- [ ] fix itest

### CODE REVIEW

File: sweep/fee_function.go [hunk 1]
```diff
@@ -86,11 +86,14 @@ type LinearFeeFunction struct {
...
+	// and the deadline block height minus one.
+	//
+	// NOTE: We do minus one here because we want to max out the budget
+	// before the deadline height is reached.
 	width uint32
 
-	// position is the number of blocks between the starting block height
-	// and the current block height.
+	// position is the fee function's current position, given a width of w,
+	// a valid position should lie in range [0, w-1].
```
Comment (morehouse) [inline sweep/fee_function.go:96]: Width now maxes out at `deadline - 1`, so we should allow position to be in [0, width].

```suggestion
	// position is the fee function's current position, given a width of w,
	// a valid position should lie in range [0, w].
```

File: sweep/fee_function.go [hunk 2]
```diff
@@ -153,18 +156,18 @@ func NewLinearFeeFunction(maxFeeRate chainfee.SatPerKWeight,
...
 	// delta is zero, it means the starting and ending fee rates are the
 	// same, which means there's nothing to increase, so any width greater
 	// than 1 doesn't provide any utility. This could happen when the
-	// sweeper is offered to sweep an input that has passed its deadline.
+	// budget is too small.
 	if l.deltaFeeRate == 0 && l.width != 1 {
 		log.Errorf("Failed to init fee function: startingFeeRate=%v, "+
 			"endingFeeRate=%v, width=%v, delta=%v", start, end,
-			confTarget, l.deltaFeeRate)
+			l.width, l.deltaFeeRate)
```
Comment (morehouse) [inline sweep/fee_function.go:170]: What is the reason for this check?  Can we remove it?

It seems like the check prevents broadcasting the transaction in cases where the budget is actually enough.  For example:

- `confTarget` is 2
- starting feerate calculated from `confTarget` is 30 sat/vB
- ending feerate for budget is 30 sat/vB

If we broadcast immediately at 30 sat/vB, the transaction is likely to confirm in time.  But currently an error is returned instead, and broadcasting is delayed until the next block when `confTarget` is 1 and the branch above is taken instead.  But we are likely to miss the deadline in that case, since we broadcasted late.

**Review by morehouse — COMMENTED** [review]
Needs an itest for #8738.

File: contractcourt/channel_arbitrator.go [hunk 3]
```diff
@@ -1535,6 +1550,11 @@ func (c *ChannelArbitrator) findCommitmentDeadlineAndValue(heightHint uint32,
 			"deadlineMinHeight=%d, heightHint=%d",
 			c.cfg.ChanPoint, deadlineMinHeight, heightHint)
 		deadline = 1
+
+	// Use half of the deadline delta, and leave the other half to be used
+	// to sweep the HTLCs.
+	default:
+		deadline /= 2
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply contractcourt/channel_arbitrator.go:1557]: updated

File: sweep/fee_function.go (hunk 1 above)
Comment (yyforyongyu) [reply sweep/fee_function.go:96]: fixed!

File: sweep/fee_function.go (hunk 2 above)
Comment (yyforyongyu) [reply sweep/fee_function.go:170]: As shown in the unit test, when conf target is 2, the max fee rate will be used. The check is added so we don't have a fee rate delta of 0, since in that case the fee func cannot do anything, also indicates the budget is too small.

File: contractcourt/channel_arbitrator.go [hunk 4]
```diff
@@ -1470,8 +1474,19 @@ func (c *ChannelArbitrator) findCommitmentDeadlineAndValue(heightHint uint32,
...
+
+		// The deadline is default to the current deadlineMinHeight,
+		// and it's overwritten when it's not none.
+		deadline := deadlineMinHeight
+		deadlineOpt.WhenSome(func(d int32) {
+			deadline = uint32(d)
+		})
+
+		if deadline < deadlineMinHeight {
+			deadlineMinHeight = deadline
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply contractcourt/channel_arbitrator.go:1489]: Yeah that's what I did initially, and @ProofOfKeags may have a different option here - but I think the idea is to do minimal work in `WhenSome`, ideally we only extract the value and quickly exit this closure.

[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply contractcourt/channel_arbitrator.go]: The easiest "formally correct" thing to do here is to use `UnwrapOr`.

```go
deadlineOpt := ...
deadline := uint32(deadlineOpt.UnwrapOr(int32(deadlineMinHeight)))
```

`WhenSome` should be considered a last resort as it is kind of an escape hatch. I included it because it makes "performing an action" with the full option value quite nice, but given that golang is an imperative language, "performing an action" very often means y'all make the choice to mutate some variable which introduces the cognitive overhead of tracking its value *over time*.

[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply contractcourt/channel_arbitrator.go]: Catching up on context it looks like this is a highly stateful algorithm where we are gradually updating `deadlineMinHeight`. My original code suggestion still works but it won't really move the needle on the overall ease of understanding the surrounding context, so feel free to ignore it if you'd rather.

File: contractcourt/channel_arbitrator.go [hunk 5]
```diff
@@ -1468,10 +1472,24 @@ func (c *ChannelArbitrator) findCommitmentDeadlineAndValue(heightHint uint32,
...
+
+		// The deadline is default to the current deadlineMinHeight,
+		// and it's overwritten when it's not none.
+		deadline := deadlineMinHeight
+		deadlineOpt.WhenSome(func(d int32) {
+			deadline = uint32(d)
+
+			// We only consider the value is under protection when
+			// it's time-sensitive.
+			totalValue += value
```
Comment (Roasbeef) [inline contractcourt/channel_arbitrator.go:1488]: Why shouldn't the total value also include this value if it isn't under a deadline? Won't they all be offered to the sweeper with a set budget, witch'll be used as fee bumping resource for the inputs it's bundled with?

File: cmd/lncli/walletrpc_active.go [hunk 6]
```diff
@@ -287,10 +287,11 @@ func bumpFee(ctx *cli.Context) error {
...
 	resp, err := client.BumpFee(ctxc, &walletrpc.BumpFeeRequest{
-		Outpoint:   protoOutPoint,
-		TargetConf: uint32(ctx.Uint64("conf_target")),
-		Immediate:  immediate,
-		Budget:     ctx.Uint64("budget"),
+		Outpoint:    protoOutPoint,
+		TargetConf:  uint32(ctx.Uint64("conf_target")),
+		Immediate:   immediate,
+		Budget:      ctx.Uint64("budget"),
+		SatPerVbyte: ctx.Uint64("sat_per_vbyte"),
```
Comment (Roasbeef) [inline cmd/lncli/walletrpc_active.go:294]: Ah must've been lost in the wash with the original version when budget replaced everything else.

File: contractcourt/chain_arbitrator.go [hunk 7]
```diff
@@ -1287,6 +1287,10 @@ func (c *ChainArbitrator) FindOutgoingHTLCDeadline(scid lnwire.ShortChannelID,
 			continue
 		}
 
+		// Make sure the channel arbitrator has the latest view of its
+		// active HTLCs.
+		channelArb.updateActiveHTLCs()
```
Comment (Roasbeef) [inline contractcourt/chain_arbitrator.go:1292]: Shouldn't this always be the case as when we revoke we update the in memory HTLC set, and the latest is read in during restart?

Comment (yyforyongyu) [reply contractcourt/chain_arbitrator.go:1292]: yeah for the channel in the process of force closing yes. But here we are reading its upstream channel, not the FCed channel, so in an edge case shown in this [test](https://github.com/lightningnetwork/lnd/pull/8751/commits/ed56a94c98127a6b705572d54fcdb8af8f2f3d1a#diff-235db90fd6236fbbc8d9ae9fd7a5d50bea961ae10d6867b90ec9a4510ba01e8eL720), the upstream channel would have no active htlcs.

I think for an active channel, we only `updateActiveHTLCs` here in `stateStep`, https://github.com/lightningnetwork/lnd/blob/87d5170dec36dba1365e7b50cc1415d9c690667b/contractcourt/channel_arbitrator.go#L935-L939

which is called inside `advanceState` and managed inside `channelAttendant`, which is an event loop that can be triggered via,
https://github.com/lightningnetwork/lnd/blob/87d5170dec36dba1365e7b50cc1415d9c690667b/contractcourt/channel_arbitrator.go#L2770
https://github.com/lightningnetwork/lnd/blob/87d5170dec36dba1365e7b50cc1415d9c690667b/contractcourt/channel_arbitrator.go#L2802
https://github.com/lightningnetwork/lnd/blob/87d5170dec36dba1365e7b50cc1415d9c690667b/contractcourt/channel_arbitrator.go#L2816
https://github.com/lightningnetwork/lnd/blob/87d5170dec36dba1365e7b50cc1415d9c690667b/contractcourt/channel_arbitrator.go#L2842
https://github.com/lightningnetwork/lnd/blob/87d5170dec36dba1365e7b50cc1415d9c690667b/contractcourt/channel_arbitrator.go#L2914
https://github.com/lightningnetwork/lnd/blob/87d5170dec36dba1365e7b50cc1415d9c690667b/contractcourt/channel_arbitrator.go#L2984
https://github.com/lightningnetwork/lnd/blob/87d5170dec36dba1365e7b50cc1415d9c690667b/contractcourt/channel_arbitrator.go#L3045
https://github.com/lightningnetwork/lnd/blob/87d5170dec36dba1365e7b50cc1415d9c690667b/contractcourt/channel_arbitrator.go#L3068

And an active channel won't update its htlcs unless a new block comes in, which is why if we mined a block in the test, it will pass.

File: contractcourt/channel_arbitrator.go (hunk 5 above)
Comment (yyforyongyu) [reply contractcourt/channel_arbitrator.go:1488]: yea this HTLC will also be offered to the sweeper, but we can be less aggressive here. The idea is, we want to CPFP the force close tx, because there are HTLCs that will timeout, so it makes sense to "borrow" some values from these HTLCs to accelerate the force close tx. However, if the HTLC is not time-sensitive, for instance, an outgoing htlc from the first hop, then there's no need to accelerate the FC tx, or, there's no need to ask this HTLC to "contribute" urgentness to the FC process.

File: cmd/lncli/walletrpc_active.go (hunk 6 above)
Comment (yyforyongyu) [reply cmd/lncli/walletrpc_active.go:294]: yeah thanks @ziggie1984 for pointing this out!

**Review by ziggie1984 — APPROVED** [review]
Had some minor comments apart from that it's LGTM

File: sweep/fee_function.go (hunk 2 above)
Comment (ziggie1984) [reply sweep/fee_function.go:170]: I think it makes sense to keep this case in place and not broadcast the sweep when we still have not reached the deadline (conftarget=2) so that more inputs can eventually increase the endFeeRate. Eventually the sweep is triggered in case the deadline is reached.

File: itest/lnd_channel_force_close_test.go [hunk 8]
```diff
@@ -417,7 +375,11 @@ func channelForceClosureTest(ht *lntest.HarnessTest,
 
 	// At this point, the CSV will expire in the next block, meaning that
 	// the output should be offered to the sweeper.
-	aliceCommit := ht.AssertNumPendingSweeps(alice, 1)[0]
+	sweeps := ht.AssertNumPendingSweeps(alice, 2)
+	aliceCommit, anchorSweep := sweeps[0], sweeps[1]
+	if aliceCommit.AmountSat < anchorSweep.AmountSat {
+		aliceCommit, anchorSweep = anchorSweep, aliceCommit
+	}
```
Comment (morehouse) [inline itest/lnd_channel_force_close_test.go:382]: s/aliceCommit/localBalanceSweep

**Review by morehouse — COMMENTED** [review]
Still have an itest failing for neutrino:

```
    harness_miner.go:217: 
        	Error Trace:	/home/runner/work/lnd/lnd/lntest/harness_miner.go:217
        	            				/home/runner/work/lnd/lnd/lntest/harness.go:1770
        	            				/home/runner/work/lnd/lnd/itest/lnd_multi-hop_test.go:2499
        	            				/home/runner/work/lnd/lnd/itest/lnd_multi-hop_test.go:152
        	Error:      	Received unexpected error:
        	            	want 1, got 2 in mempool: [08f01a438347fca122b815936cfa9417a66fbbaf2d3b02497eec7c4c382ca4aa 6b76702b74421fdc0a7b0fbdb12bd076bb7ff215dac9fdaae79ab5f54655355b]
        	Test:       	TestLightningNetworkDaemon/tranche03/134-of-155/neutrino/htlc_timeout_resolver_extract_preimage_remote/zeroconf=true/committype=ANCHORS
        	Messages:   	assert tx in mempool timeout
```

And possibly related failures on windows.

File: contractcourt/channel_arbitrator.go [hunk 9]
```diff
@@ -1535,6 +1553,11 @@ func (c *ChannelArbitrator) findCommitmentDeadlineAndValue(heightHint uint32,
 			"deadlineMinHeight=%d, heightHint=%d",
 			c.cfg.ChanPoint, deadlineMinHeight, heightHint)
 		deadline = 1
+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply contractcourt/channel_arbitrator.go:1556]: When there's no htlcs on the commit `deadlineMinHeight` will be this value as it's defined above, `deadlineMinHeight := uint32(math.MaxUint32)`.

File: contractcourt/channel_arbitrator.go [hunk 10]
```diff
@@ -1535,6 +1553,11 @@ func (c *ChannelArbitrator) findCommitmentDeadlineAndValue(heightHint uint32,
...
+	// Use half of the deadline delta, and leave the other half to be used
+	// to sweep the HTLCs.
+	default:
+		deadline = (deadlineMinHeight - heightHint) / 2
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply contractcourt/channel_arbitrator.go:1560]: updated

File: sweep/fee_function.go [hunk 11]
```diff
@@ -86,11 +86,14 @@ type LinearFeeFunction struct {
 	currentFeeRate chainfee.SatPerKWeight
 
 	// width is the number of blocks between the starting block height
-	// and the deadline block height.
+	// and the deadline block height minus one.
+	//
+	// NOTE: We do minus one here because we want to max out the budget
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply sweep/fee_function.go:91]: updated

File: sweep/fee_function.go (hunk 2 above)
Comment (yyforyongyu) [reply sweep/fee_function.go:170]: > In this case, shouldn't we just broadcast with max budget immediately and hope for the best?

I guess the question is where is the hope coming from:)

I don't know if the age of a tx is still a factor for the miners these days. If not and, if the tx gets confirmed at some blocks later, it means the network fee rate goes down -> the fee func will be created then?

The initial intention is to return an error here, so the user can increase the budget - for instance, when the htlc is large and the budget ratio is small, the user can intervene by specifying a larger budget via `BumpFee`.

> It is very unlikely that future inputs will have the same deadline as the current inputs, so they are unlikely to be batched together.

Yeah this is an issue IMO - if the deadlines are already large, I think they can be grouped - like, how different in terms of urgency they have between deadlines of 1008 and 1007 - think there should be a ladder structure, sth like the fee rate bucket we used to have.

File: itest/lnd_sweep_test.go [hunk 12]
```diff
@@ -152,167 +189,513 @@ func testSweepAnchorCPFPLocalForceClose(ht *lntest.HarnessTest) {
...
+
+	// Calculate the fee and fee rate of Bob's sweeping tx.
 	fee := uint64(ht.CalculateTxFee(sweepTx))
 	feeRate := uint64(ht.CalculateTxFeeRate(sweepTx))
 
-	// Alice should start with the initial fee rate of 2000 sat/kw.
-	startFee := startFeeRate.FeeForWeight(txWeight)
+	// feeFuncWidth is the width of the fee function. By the time we got
+	// here, we've already mined one block, plus we need to confirm before
+	// deadline-1, so the width is anchor's deadline minus 2.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply itest/lnd_sweep_test.go:233]: updated thanks!

File: itest/lnd_sweep_test.go [hunk 13]
```diff
@@ -152,167 +189,513 @@ func testSweepAnchorCPFPLocalForceClose(ht *lntest.HarnessTest) {
...
+	ht.FundCoins(btcutil.SatoshiPerBitcoin, bob)
+
+	// For neutrino backend, we need one more UTXO for Bob to create his
+	// sweeping txns.
+	if ht.IsNeutrinoBackend() {
+		ht.FundCoins(btcutil.SatoshiPerBitcoin, bob)
+	}
+
+	// Subscribe the invoice.
+	stream := carol.RPC.SubscribeSingleInvoice(payHash[:])
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply itest/lnd_sweep_test.go:452]: updated

File: itest/lnd_sweep_test.go [hunk 14]
```diff
@@ -152,167 +189,513 @@ func testSweepAnchorCPFPLocalForceClose(ht *lntest.HarnessTest) {
...
+	// Mine a block so Bob's force closing tx stays in the mempool, which
+	// also triggers the CPFP anchor sweep.
+	ht.MineEmptyBlocks(1)
+
+	// Bob should still have two pending sweeps,
+	// - anchor sweeping from his local commitment.
+	// - anchor sweeping from his remote commitment (invalid).
+	ht.AssertNumPendingSweeps(bob, 2)
+
+	// We now check the expected fee and fee rate are used for Bob's anchor
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply itest/lnd_sweep_test.go:551]: hmm think the original one is correct...?

File: itest/lnd_sweep_test.go [hunk 15]
```diff
@@ -24,126 +24,163 @@ import (
...
-//  1. Fund Alice with 2 UTXOs - she will need two to sweep her anchors from
-//     the local and remote commitments, with one of them being invalid.
-//  2. Fund Bob with no UTXOs - his sweeping txns don't need wallet utxos as he
-//     doesn't need to sweep any time-sensitive outputs.
-//  3. Alice opens a channel with Bob, and sends him an HTLC without being
-//     settled - we achieve this by letting Bob hold the preimage, which means
-//     he will consider his incoming HTLC has no preimage.
-//  4. Alice force closes the channel.
+//  1. Fund Alice with 1 UTXOs - she only needs one for the funding process,
+//  2. Fund Bob with 2 UTXOs - he needs one for the funding process, one for
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply itest/lnd_sweep_test.go:33]: Yeah it's fixed! Here's what happened, since we are creating two sweeping txns for each of the anchors,
- if remote anchor gets created first -> `testmempoolaccept` says no -> the wallet utxo is unlocked -> the same wallet utxo is now used for the local anchor sweep.
- if the local anchor gets created first -> `testmempoolaccept` says yes -> the wallet utxo is locked -> when creating the remote anchor it will return `ErrNotEnoughInputs`.

File: itest/lnd_sweep_test.go [hunk 16]
```diff
@@ -152,167 +189,513 @@ func testSweepAnchorCPFPLocalForceClose(ht *lntest.HarnessTest) {
...
+	// The above mined block should confirm Bob's force close tx, and his
+	// contractcourt will offer the HTLC to his sweeper. We are not testing
+	// the HTLC sweeping behaviors so we just perform a simple check and
+	// exit the test.
+	ht.AssertNumPendingSweeps(bob, 1)
+
+	// Finally, clean the mempool for the next test.
+	ht.CleanShutDown()
+}
+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply itest/lnd_sweep_test.go:363]: I think we can make the `for` loop here into a function and change it for every fee function - tho I'd prefer every time we add a new fee func we add new tests, so it's easier to maintain. I kinda like the strictness in this test tho, makes me feel we are harnessing the behavior here.

File: itest/lnd_multi-hop_test.go [hunk 17]
```diff
@@ -713,9 +713,11 @@ func runMultiHopLocalForceCloseOnChainHtlcTimeout(ht *lntest.HarnessTest,
 	// to be mined to trigger a force close later on.
 	var blocksMined uint32
 
-	// Increase the fee estimate so that the following force close tx will
-	// be cpfp'ed.
-	ht.SetFeeEstimate(30000)
+	// We need to mine a block otherwise `FindOutgoingHTLCDeadline` cannot
+	// find the incoming HTLC on Bob's ChainArbitrator.
+	//
+	// TODO(yy): Investigate and fix it!
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply itest/lnd_multi-hop_test.go]: yeah fixed the a following commit

File: itest/lnd_channel_force_close_test.go (hunk 8 above)
Comment (yyforyongyu) [reply itest/lnd_channel_force_close_test.go:382]: changed to `commitSweep`

File: itest/lnd_channel_force_close_test.go [hunk 18]
```diff
@@ -417,7 +375,11 @@ func channelForceClosureTest(ht *lntest.HarnessTest,
 
 	// At this point, the CSV will expire in the next block, meaning that
 	// the output should be offered to the sweeper.
-	aliceCommit := ht.AssertNumPendingSweeps(alice, 1)[0]
+	sweeps := ht.AssertNumPendingSweeps(alice, 2)
+	aliceCommit, anchorSweep := sweeps[0], sweeps[1]
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply itest/lnd_channel_force_close_test.go:379]: nope - we do a check below to find the right sweep.

**Comment by yyforyongyu** [comment]
@morehouse Yeah the itest failure is annoying, don't think it can be properly fixed until `blockbeat` is in. Here's some logs which show that the subsystems have very different views of block heights,
```
2024-05-18 04:24:15.394 [DBG] SWPR: Received new block: height=476, attempt sweeping 0 inputs
...
2024-05-18 04:24:15.418 [INF] CNCT: *contractcourt.htlcSuccessResolver(ee04c77ad329073b6b7e1899e0173aa6c1835727f647b0853099fc7dcc6e0d2b): offering second-level success tx output to sweeper with no deadline and budget=0.00050000 BTC at height=471
...
2024-05-18 04:24:15.418 [INF] SWPR: Sweep request received: out_point=e0749b9458c5f6693eb79b6a6293feb44435467b2a68680185921f0d86425805:0, witness_type=HtlcAcceptedSuccessSecondLevel, relative_time_lock=4, absolute_time_lock=0, amount=0.00100000 BTC, parent=(<nil>), params=(startingFeeRate={false 0}, immediate=false, exclusive_group=none, budget=0.00050000 BTC, deadline=none)
2024-05-18 04:24:15.418 [INF] CNCT: *contractcourt.htlcSuccessResolver(ee04c77ad329073b6b7e1899e0173aa6c1835727f647b0853099fc7dcc6e0d2b): waiting for second-level HTLC output to be spent after csv_delay=4
...
2024-05-18 04:24:15.418 [DBG] SWPR: TxPublisher received new block: 471
```

As for now, I did an extra check in the test to make it pass. More work to do in 18.1, as I wanna fix these multi-hop tests so it's easier to maintain in the future.

File: itest/lnd_sweep_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply itest/lnd_sweep_test.go:360]: If/when we add other fee bumping functions (open questions of when those would be used, how we'd test compare, etc, etc), we can abstract the test further to allow for fee function specific assertions.

**Review by Roasbeef — APPROVED** [review]
LGTM 🐠

File: sweep/fee_function.go (hunk 2 above)
Comment (morehouse) [reply sweep/fee_function.go:170]: > I don't know if the age of a tx is still a factor for the miners these days. If not and, if the tx gets confirmed at some blocks later, it means the network fee rate goes down -> the fee func will be created then?

Maybe the fee func will be created then, or maybe not.  How sensitive is bitcoind's fee estimator to temporary drops in fee rate?  If not too sensitive, then the fee func will continue to fail.

And bitcoind's fee estimator is *backwards-looking*, so it will always be delayed in adjusting downward.  So by the time we actually broadcast the transaction, it might be too late (because deadline is missed or because we missed the lower-fee window and mempool fees went back up).

Overall, I think we're better off to have the transaction in the mempool so it can potentially be mined.  Until we broadcast it, it can *never* be mined.

> The initial intention is to return an error here, so the user can increase the budget - for instance, when the htlc is large and the budget ratio is small, the user can intervene by specifying a larger budget via `BumpFee`.

Perhaps an early sanity check specifically for `BumpFee` is more appropriate?  If budget is less than bitcoind's estimate, we can quickly return an error from `BumpFee` instead of processing further.

For contractcourt sweeps, I think it's better to broadcast at max budget than to give up completely.

> Yeah this is an issue IMO - if the deadlines are already large, I think they can be grouped - like, how different in terms of urgency they have between deadlines of 1008 and 1007 - think there should be a ladder structure, sth like the fee rate bucket we used to have.

+1
